src/librustc_mir/transform/const_prop.rs

```diff
@@ -768,6 +772,8 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
 enum ConstPropMode {
     /// The `Local` can be propagated into and reads of this `Local` can also be propagated.
     FullConstProp,
+    /// The `Local` can only be propagated into and from its own block.
+    OnlyInsideOwnBlock,
```

Comment:
Ah interesting, my original idea was to only propagate locals that are written to and read exactly once in the entire MIR, but this does the job, too, and is even more powerful in general.